osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java

```diff
@@ -0,0 +1,233 @@
+		if (pMilestoneManagers == null) {
+			if (mMilestoneManagers.size() > 0) {
+				mMilestoneManagers.clear();
+			}
+		} else {
+			mMilestoneManagers = pMilestoneManagers;
+		}
+	}
+
+	/**
+	 * @return aggregate distance (in meters)
+	 */
+	public double getDistance() {
+		return mOutline.getDistance();
+	}
+
+	/** Internal method used to ensure that the infowindow will have a default position in all cases,
+	 * so that the user can call showInfoWindow even if no tap occured before.
+	 * Currently, set the position on the center of the polygon bounding box.
+	 */
+	protected void setDefaultInfoWindowLocation() {
+		int s = mOutline.getPoints().size();
+		if (s == 0){
+			mInfoWindowLocation = new GeoPoint(0.0, 0.0);
+			return;
+		}
+		//TODO: as soon as the polygon bounding box will be a class member, don't compute it again here.
+		mInfoWindowLocation = mOutline.getCenter(null);
+	}
+
+	@Override
+	public void draw(final Canvas pCanvas, final Projection pProjection) {
+		if (mPath != null) {
+			drawWithPath(pCanvas, pProjection);
+		} else {
+			drawWithLines(pCanvas, pProjection);
+		}
+	}
+
+	private void drawWithPath(final Canvas pCanvas, final Projection pProjection) {
+		mPath.rewind();
+
+		mOutline.setClipArea(pProjection);
+		final PointL offset = mOutline.buildPathPortion(pProjection, null, mMilestoneManagers.size() > 0);
+		for (final MilestoneManager milestoneManager : mMilestoneManagers) {
+			milestoneManager.init();
+			milestoneManager.setDistances(mOutline.getDistances());
+			for (final PointL point : mOutline.getPointsForMilestones()) {
+				milestoneManager.add(point.x, point.y);
+			}
+			milestoneManager.end();
+		}
+
+		if (mHoles != null) {
+			for (final LinearRing hole : mHoles) {
+				hole.setClipArea(pProjection);
+				hole.buildPathPortion(pProjection, offset, mMilestoneManagers.size() > 0);
+			}
+			mPath.setFillType(Path.FillType.EVEN_ODD); //for correct support of holes
+		}
+
+		if (mFillPaint != null) {
+			pCanvas.drawPath(mPath, mFillPaint);
+		}
+		pCanvas.drawPath(mPath, mOutlinePaint);
+
+		for (final MilestoneManager milestoneManager : mMilestoneManagers) {
+			milestoneManager.draw(pCanvas);
+		}
+
+		if (isInfoWindowOpen() && mInfoWindow!=null && mInfoWindow.getRelatedObject() == this) {
+			mInfoWindow.draw();
+		}
+	}
+
+	private void drawWithLines(final Canvas pCanvas, final Projection pProjection) {
+		mLineDrawer.setCanvas(pCanvas);
+		mOutline.setClipArea(pProjection);
+		mOutline.buildLinePortion(pProjection, mMilestoneManagers.size() > 0);
+		for (final MilestoneManager milestoneManager : mMilestoneManagers) {
+			milestoneManager.init();
+			milestoneManager.setDistances(mOutline.getDistances());
+			for (final PointL point : mOutline.getPointsForMilestones()) {
+				milestoneManager.add(point.x, point.y);
+			}
+			milestoneManager.end();
+		}
+
+		for (final MilestoneManager milestoneManager : mMilestoneManagers) {
+			milestoneManager.draw(pCanvas);
```

Comment:
Can the draw be done in the previous loop after the milestone manager is populated, or does it cause delayed draw behavior that way?

Reply:
We need these two steps.
The first step computes the actual "milestones", which can be something like a milestone every km, or 5 equidistant milestones for the whole length of the polyline. Therefore we need to parse everything first. Then draw, in the second step.